
# HG changeset patch
# User Todd C. Miller <Todd.Miller@courtesan.com>
# Date 1462373721 21600
# Node ID 1d13341d53ecefb58cfac03374a66aedd9c2e1b4
# Parent  910af8ba46664e7c8a94402f02cb01c03533a771
Sync the "cache hit" debug messages with the "cached" debug messages.
This fixes a bug where we could dereference a NULL pointer when we
look up a negative cached entry which is stored as a NULL passwd
or group struct pointer.  Bug #743.

diff -r 910af8ba4666 -r 1d13341d53ec plugins/sudoers/pwutil.c
--- a/plugins/sudoers/pwutil.c	Thu Apr 28 15:27:53 2016 -0600
+++ b/plugins/sudoers/pwutil.c	Wed May 04 08:55:21 2016 -0600
@@ -139,8 +139,8 @@
 	item = node->data;
 	sudo_debug_printf(SUDO_DEBUG_DEBUG,
 	    "%s: uid %u [%s] -> user %s [%s] (cache hit)", __func__,
-	    (unsigned int)uid, key.registry, item->d.pw->pw_name,
-	    item->registry);
+	    (unsigned int)uid, key.registry,
+	    item->d.pw ? item->d.pw->pw_name : "unknown", item->registry);
 	goto done;
     }
     /*
@@ -202,8 +202,8 @@
     if ((node = rbfind(pwcache_byname, &key)) != NULL) {
 	item = node->data;
 	sudo_debug_printf(SUDO_DEBUG_DEBUG,
-	    "%s: user %s [%s] -> uid %u [%s] (cache hit)", __func__, name,
-	    key.registry, (unsigned int)item->d.pw->pw_uid, item->registry);
+	    "%s: user %s [%s] -> uid %d [%s] (cache hit)", __func__, name,
+	    key.registry, item->d.pw ? (int)item->d.pw->pw_uid : -1, item->registry);
 	goto done;
     }
     /*
@@ -461,8 +461,8 @@
 	item = node->data;
 	sudo_debug_printf(SUDO_DEBUG_DEBUG,
 	    "%s: gid %u [%s] -> group %s [%s] (cache hit)", __func__,
-	    (unsigned int)gid, key.registry, item->d.gr->gr_name,
-	    item->registry);
+	    (unsigned int)gid, key.registry,
+	    item->d.gr ? item->d.gr->gr_name : "unknown", item->registry);
 	goto done;
     }
     /*

